docs/proposals/unified-backend-architecture/tech-spec.md

```diff
@@ -0,0 +1,221 @@
+for example, UTXO.
+
+## Assumptions
+
+N/A, see **Prerequisites**.
+
+## Requirements
+
+Unification of the wallet backend architecture should allow to:
+
+1. work with different hardware wallets in the same manner,
+2. create lightweight Daedalus mobile clients (iOS/Android),
+3. create lightweight Daedalus desktop clients (Windows/macOS/Linux).
+
+The main benefit of such a lightweight client is independence from the growing blockchain: the client does not
+store blockchain locally and does not wait for a "Syncing blocks".
+
+## Key Idea
+
+As mentioned before, the key idea is to unify the different use cases under a single architecture, we mean
+hardware wallets, mobile clients and lightweight desktop clients.
+
+In the case of [Ledger](https://www.ledgerwallet.com/) (or any other hardware wallet) the privte keys and the
+cryptography are hosted on the device, whereas in a mobile client they are held on a smartphone. And this schema
+fits for "featherweight Daedalus desktops" that do not want to host the growing blockchain on the computer's
+storage.
+
+More specifically, a Daedalus mobile client will consist of an iOS or Android application which will communicate
+with a cluster of nodes that will keep the global UTXO state for the blockchain and that will be capable
+to reconstruct each mobile wallet state without hosting any user-specific data.
+
+## Multi-Tenant Architecture
+
+Mobile clients will communicate with a clusters of nodes, so the basic requirement is an ability of one node to work
+with particular number of clients simultaneously. For example, if 1 cluster has to serve 10000 mobile clients and
+this cluster consists of 100 nodes, each node has to be able to work with 100 clients simultaneously and reliably.
+Thus, the longest procedure is a new payment (creating new transaction), and even if all these 100 clients will
+sign their transactions and send it for publishing at the same time, the node must be able to handle these
+transactions reliably and as fast as possible.
+
+To achieve this goal the node must implement robust concurrent model. More specifically, all requests to the wallet
+web API endpoints must be handled concurrently.
+
+**TODO**: To think about the slowest endpoints, like `/api/v1/transactions`: is it possible to handle them
+fully concurrently? I mean publishing to the network, storing in the `wallet-db`, etc.
+
+## Scalability
+
+Initially clusters of nodes will be hosted by IOHK (on AWS EC2). And we need a mechanism of auto-scaling of the
+cluster: it must be able to grow proportionally to growing number of clients.
+
+**TODO**: To think about auto-scale supervisor. Suppose we have 1 cluster that consists of 100 nodes and each of node
+can reliably serve 100 clients. If supervisor sees that number of clients connected to this cluster is growing (and
+becomes more than 10000), it launches additional nodes.
```

Comment:
@domenkozar What do you think about it?

Reply:
We might not base our autoscaling on number of users, but another resource. It's hard to say how it would scale before we run some benchmarks. It could be that bottleneck is CPU and you can't correlate that with number of subscribers.

Reply:
Right, exact resource and tuning to do to be decided during detailed design analysis or later benchmarking.